exposed utils namespace

tomatau opened the pull request:
Exposed all utils into utils namespace.

fixes https://github.com/react-bootstrap/react-bootstrap/issues/743

Kept the reference to `EventListener` until a third party option is used instead.

Mapped Object.assign.js to just `assign`

mtscout6 commented on src/utils/index.js:
```diff
@@ -0,0 +1,21 @@
+  CustomPropTypes,
+  deprecationWarning,
+  domUtils,
+  EventListener,
```
I'd rather not have this one here, since there is one available in React. So, long as you are using npm and CommonJS or ES6 Modules.

tomatau replied:
If EventListener is being removed from this namespace I think it should be removed from the libraries internals entirely and specified as a dependency.

Object.assign should be included via the ES6 -> ES5 build step.

mtscout6 replied:
That is the goal in the long run, but removing an internal usage is easier than removing something with explicit public exposure.

mtscout6 commented on src/utils/index.js:
```diff
@@ -0,0 +1,21 @@
+  deprecationWarning,
+  domUtils,
+  EventListener,
+  assign,
```
Same applies to this one.

taion commented:
I want to limit the line notes, but it seems odd to me to expose `deprecationWarning`, and I'd strongly prefer that `createContextWrapper` not be exposed because it's not meant to be public.

Maybe instead of taking this approach, we should only expose `domUtils`, and add others as it becomes relevant?

tomatau commented:
I agree to limit exposure and deprecationWarning doesn't feel suitable to expose.

I've already found use for `domUtils`, `createChainedFunction` and `EventListener` when building a custom overlay trigger.

Also, it felt far more suitable to use this libs EventListener than duplicate the functionality of a cross browser event listener in my codebase. I feel this should be moved elsewhere.

Also `CustomPropTypes` seems like it would be popular.

mtscout6 commented:
The `EventListener` is available today from the React npm module. Just use:

```
import EventListener from 'react/lib/EventListener';
```

taion commented:
This is in the context of using the bundles, no? Is `EventListener` exposed in ReactWithAddons?

Although I think I'm still inclined to say that if you are using the bundle, chances are you're not going to want/need to deal with this sort of thing.

mtscout6 commented:
No `EventListener` is not available on ReactWithAddons, would be nice if it was. With that said though it is still there, and if they remove it then we can revisit the idea of including it in this library or another one.

tomatau commented:
I wasn't aware it was available through the lib :+1:

mtscout6 commented:
Can you squash those commits and use the commit message format outlined in the [Contributing Guide](https://github.com/react-bootstrap/react-bootstrap/blob/master/CONTRIBUTING.md#commit-subjects-for-public-api-changes). That will allow your change to be reflected in our generated changelog on the next release.

taion commented on src/utils/index.js:
```diff
@@ -0,0 +1,15 @@
+export default {
+  childrenValueInputValidation,
+  createChainedFunction,
+  createContextWrapper,
```
I'd really rather this not be exposed. There's intentionally no user-facing documentation for any internal users of this function.

taion commented:
One more thought - both this and `src/index.js` would be a lot cleaner if we enabled `es7.exportExtensions` under Babel Stage 1. You'd just write this whole thing as

```js
export childrenValueInputValidation from './childrenValueInputValidation';
export createChainedFunction from './createChainedFunction';
export CustomPropTypes from './CustomPropTypes';
export domUtils from './domUtils';
export ValidComponentChildren from './ValidComponentChildren';
```

Probably not necessary here, but it'd make things cleaner as a separate change.

tomatau commented:
squashed and added a commit prefix

NB. It's my first time squashing - is it supposed to persist the previous commit messages? The squash claims to be successful but I can still see them in the log and this PR history...

mathieumg commented:
> is it supposed to persist the previous commit messages

Not if you squashed them together with a new message. Did you force-push, or do you still see them in your local log?

tomatau commented:
@mathieumg 

I reset -i HEAD~4, picked one and squashed 3, changed the commit message of the one I picked to be more suitable, saved and wrote.. then was prompted for another commit message so repeated the previous one.

At this point my local log was only showing 1 commit.

After this - I realised that I'd made these changes onto my local master - so branched out before pulling origin master (on my fork) then merged my squashed changes back in and pushed.

At this point my local log was showing all commits :palm_tree:

dozoisch commented:
Basically

```
git rebase -i HEAD~X
```

Then you can do pick, s, s, s, s, etc. (f to squahs ignoring the commit message)

Then

```
git push --force origin branch_name
```

You have to push force with a squash because you are rewriting history.

mtscout6 commented:
If you haven't seen this before it's a great video highlighting many of the cool things you can do with git. Squashing is one of the topics covered.

https://vimeo.com/43659036

AlexKVal commented:
> Also CustomPropTypes seems like it would be popular.

It is already the second voice for it (the first was in Gitter).

npm search `react-custom-prop-types` is free :wink:

mtscout6 commented:
Other available names:

- `react-prop-types` (I'm partially in favor of this one)
- `react-enhanced-prop-types`
- `react-prop-type-utils`

mathieumg commented:
If `react-prop-types` is free, why not go with it!

tomatau commented:
could the chat about custom-prop-types be moved to another issue?

AlexKVal commented:
Yeah. I am sorry about starting it here :cherries: 
We should continue it in gitter for maintainers.

mtscout6 commented:
@tomatau Sorry about that, will do

Did that video help you understand how to squash commits a little easier?

tomatau commented:
@mtscout6 thanks for the video - didn't add anything I didn't know about squash (something more complicated broke my push) but it did have a lot of really nice tricks and advice that I'm very grateful for!

I'm going to close this PR and submit a new one with a cleaner commit history
